### ISSUE TITLE
Admin page: export form definitions and data with attachments

### ISSUE THREAD

**Issue by ebruchez** [orbeon/orbeon-forms#779]
## Feature

The form definition, its dependencies (images, attachments), and/or form data and dependencies, are exported into a zip archive, which can be imported back into a Form Builder/Form Runner installation.
## Use cases
- Forms are developed on a instance of Form Builder distinct from the one used for staging and/or production and not reachable via direct internet connection.
- Archival/backup of forms
- Exchange of forms
- ~~Migrating from one persistence implementation to another (e.g. eXist to relational)~~
    - We won't support exporting from eXist at this point as that implementation has been removed.
- Exporting from database to Orbeon Forms resources
## Location of functionality
- Form Runner Admin page
- Form Runner Summary page
- P2: Form Builder Summary/Detail pages for non-published form definitions
## Steps
- retrieve form definition, related dependencies, form data
- use zip processor (or probably from Scala) to produce downloadable zip file
- multiple forms can be selected for export when possible
- specific data can be selected for export when possible (from Summary page only)
## Service

It might be good to implement most of the functionality as a service first. This would also enable a scripted mode which would not require the Orbeon Forms user interface.

_NOTE: This however would still require a running Form Builder instance._

- This is implemented as a service that returns a zip archive or an error.
- You can pass the service the id of a form, and/or search values as shown in the Form Builder Summary page. This selects 0 or more forms.
- Each form selected is included in the archive.
- If possible, calling the service from the command-line could be done with a simple wrapper around a tool like  - wget/curl. If that is not possible or desirable, another solution can be found.

**Comment by ebruchez** [comment]
Also importing/exporting data with dependencies should also be an option.

**Comment by ebruchez** [comment]
Also requested: ability to export selected subset of data only. See [Customer thread](https://basecamp.com/1721271/projects/5115898/messages/37771772#comment_252875058).

**Comment by ebruchez** [comment]
[+1 from customer](https://basecamp.com/1721271/projects/13032841/messages/63038828#comment_468042384)

**Comment by ebruchez** [comment]
[+1 from customer](https://3.basecamp.com/3600924/buckets/1932446/messages/1052091828)

**Comment by ebruchez** [comment]
For this [customer ](https://3.basecamp.com/3600924/buckets/1932446/messages/1052091828#__recording_1077419891), handling form definition only is enough and there is no requirement to handle data as well. This might simply at little bit a first implementation.

**Comment by ebruchez** [comment]
Versioning should be handled as well:

- upon exporting, the user must be able to select the version(s) to export
- version numbers must be included in the zip archive
- import must provide options to select
  - which items to import
  - how to map versions
  - whether to make form definitions available or not

See also  #1451 and #3448 about relevant improvements to versioning on the Home page.

**Comment by ebruchez** [comment]
The requirement also calls for:

- doing this manually via a Form Runner UI
- doing this via an API

**Comment by ebruchez** [comment]
[+1 from customer](https://3.basecamp.com/3600924/buckets/4838999/messages/1172186533#__recording_1174132042) as an option to migrate between providers although another, explicit way of migrating between  providers might be better).

**Comment by ebruchez** [comment]
[+1 from customer](https://mail.google.com/mail/u/0/#inbox/FMfcgxwHNVzNsxXbdJrhRwfFgjVhWDGZ)

**Comment by ebruchez** [comment]
+1 from customer for archival of data.

[ZIP64](https://en.wikipedia.org/wiki/ZIP_(file_format)#ZIP64) came out with the ZIP 4.5 specification in 2001 and Java 7 (2011) supports ZIP64. So file size for large exports should not be a problem.

**Comment by ebruchez** [comment]
First step: write "archive export API". Parameters:

- date range
    - all
    - `from-last-modified-time`
    - `until-last-modified-time` (optional)
- data vs. form vs. both (vs. attachments?)
    - `content=form-definition|data|attachments` ?
- app/form/form-version
    - app: part of the path, optional
    - form: part of the path, optional
    - `form-version`: URL parameter or header?
        -  [Search API](https://doc.orbeon.com/form-runner/api/persistence/search#versioning) uses `Orbeon-Form-Definition-Version` header for latest/specific/all
            - but that might not have been the best choice vs. a URL parameter
    - `document-id`: part of the path, optional (requires app/form)
        - scenario: export a single instance of form data with its attachments

- [ ] P2: "dry run" to just return the count(s) of operations

New entry point:

```
/fr/service/persistence/export/$app/$form/$document-id?form-version=$form-version
```

`$app`/`$form`/`$document-id`/`form-version` are all optional, but in constrained ways:

- no parameter: export everything
    - we can imagine allowing `form-version=latest|all`, but `specific` doesn't make much sense here
- `$app` only: export everything under the app
    - we can imagine allowing `form-version=latest|all`, but `specific` doesn't make much sense here
- `$app` and `$form`
    - `form-version=latest|specific|all`
- `$app` and `$form` and `document-id`
    - `form-version`: can be omitted or specific, but if so must match

**Comment by ebruchez** [comment]
We are implementing this mostly at the level of the persistence proxy, not directly in the providers.

- a request that needs to search for form definitions and their versions, use the [Form metadata API](https://doc.orbeon.com/form-runner/api/persistence/forms-metadata)
- a request that needs to search for data uses the [Search API](https://doc.orbeon.com/form-runner/api/persistence/search#versioning)
- `GET`ting data and attachments use the [CRUD API](https://doc.orbeon.com/form-runner/api/persistence/crud)

**Comment by ebruchez** [comment]
- [ ] export form definitions with attachments only
    - [ ] support app/form/version filters
- [ ] export form data with attachments only
    - [ ] call Search API
    - [ ] retrieve form data and attachments

**Comment by ebruchez** [comment]
Clarification: for the date range filtering, we probably want this to apply to the data, not the form definition. (If we would like some filtering applying to form definitions, we need to be explicit and add other parameters.)

For app/form/version triplet selection, we have:

- all forms
- filtering by range
- specific triplets

If we pass lots of triplets, we need to watch out for:

- how to encode them
- URL length

However, internally, we don't have a URL length limit, and the use case is the Admin page passing a large number of triplets to the API. We are ok passing this as URL parameters for now, and later, if needed, we can add a `POST` option.

**Comment by ebruchez** [comment]
For date/times, we need clarity. Suggesting (updated above):

- `last-modified-time-gt`/`last-modified-time-ge`
- `last-modified-time-lt`/`last-modified-time-le` (optional)

**Comment by ebruchez** [comment]
Realizing that there is a conflict between:

- the ability to specify app/form/document-id and form version by path and URL parameter (for the form version)
- the ability to pass a series of triplets via URL parameter

Should we support both? The Admin page will support the triplets, since that is the result of the selection in the UI.

**Comment by ebruchez** [comment]
For how attachments are currently handled by Form Runner, see [this comment](https://github.com/orbeon/orbeon-forms/issues/4146#issuecomment-1549987193). This said, at the API/service level/database, we do support updates to attachments.

In theory, we could call the Data History on attachments as well.

Right now, we don't. Scenario:

- save form data with attachments
- don't change attachments but save data again

Result:

```
issue/779/1/form/latest/form.xhtml
issue/779/1/data/9e217967755f608bd8e90b194a88b64482dc74b9/2023-06-14T20:48:02Z/data.xml
issue/779/1/data/9e217967755f608bd8e90b194a88b64482dc74b9/2023-06-14T20:48:02Z/edd1cbc2f97dd5240303de095e30cda779d1e8ef.bin
issue/779/1/data/9e217967755f608bd8e90b194a88b64482dc74b9/2023-06-14T20:48:02Z/c0d22317677b5ea07885d710e0a0a6318a94a779.bin
issue/779/1/data/9e217967755f608bd8e90b194a88b64482dc74b9/2023-06-14T20:47:11.941Z/data.xml
issue/779/1/data/9e217967755f608bd8e90b194a88b64482dc74b9/2023-06-14T20:47:11.941Z/edd1cbc2f97dd5240303de095e30cda779d1e8ef.bin
issue/779/1/data/9e217967755f608bd8e90b194a88b64482dc74b9/2023-06-14T20:47:11.941Z/c0d22317677b5ea07885d710e0a0a6318a94a779.bin
```

This is not good because the attachments are duplicated.

**Comment by ebruchez** [comment]
Solution: 

- for a given document id, keep track of attachment filenames
- write the first one as `latest`
- duplicate attachment references are ignored

**Comment by ebruchez** [comment]
For data date ranges:

- `include-data-revision-history=false`
    - simply check the current data's date is within range
- `include-data-revision-history=true`
    - include all the entries within the range
    - this means current data might not be included
    - if data is deleted, nothing will be exported (see above: we need change to Search API to support that)

**Comment by ebruchez** [comment]
The UI must allow controlling the following axes:

- form vs. data
    - form definitions only
    - form data only
    -  both
- current data vs. revision history
    - current data only
    - also include revision history
    - no need to allow for "revision history only"
- app/form
    - everything
    - everything for a given app
    - everything for a given app/form
    - from selection (checkboxes)
- form versions
    - latest only
    - all
    - from selection (checkboxes)
- date ranges
    - only data older than x
    - only data newer than x

All of these options require a dialog. We must have a button or an entry in the Operations menu to open that "Export" dialog. Once the options are selected, an "Export" (vs. "Cancel") button will start the export and download the zip.

**Comment by ebruchez** [comment]
For timestamps, we get them through the following APIs:

- CRUD
- Search
- History

In the first case, we use the standard `Created`/`Last-Modified` headers, and those only have a resolution of one second. IN the other cases, we return our own ISO dates which have millisecond resolution. Suggesting adding millisecond ISO dates to the `GET`/`HEAD` calls as well. If present, those will be used for the purpose of setting the last modification dates in the Zip archive entry metadata and paths.

**Comment by ebruchez** [comment]
- [ ] the autocomplete fields in the dialog do not allow entering a search text

<img width="964" alt="image" src="https://github.com/orbeon/orbeon-forms/assets/105769/b8dce57c-2f17-489f-8b48-3cdf5c886d37">

**Comment by ebruchez** [comment]
- [ ] forward-port changes to `master` and custom 2022.1-pe branch @ebruchez

**Comment by ebruchez** [comment]
Created separate issue #5924 for import.

**Comment by ebruchez** [comment]
- [ ] check that `form-version=all` works (didn't work today for me)

**Comment by ebruchez** [comment]
- [ ] consider optimizing export/purge of historical data by running a single SQL query instead of issuing individual `DELETE` or `GET` (either implement, or enter new RFE)

**Comment by ebruchez** [comment]
- [ ] consider #6023

**Comment by ebruchez** [comment]
- [ ] doc (@obruchez ?)

**Comment by ebruchez** [comment]
Permissions: I think we had decided that this was a admin function, which would export all data regardless of permissions.

- [ ] check whether this is indeed the current behavior
- [ ] document that aspect

**Comment by ebruchez** [comment]
To check the permissions behavior:

- [ ] as user `clerk`, enter data that only user can read, with no permissions for user `admin`
- [ ] as user `admin`, export form and data

**Comment by ebruchez** [comment]
Conclusion of the above:

- as `admin` I am getting a zip with the form definition, but no data
- as `clerk` (which also can access the Admin page), I am not getting any download (and no error), but there is a 403 in the logs

**Comment by ebruchez** [comment]
The 403 comes when:

- calling `/fr/service/mysql/history/test/export-permissions/2cadd6415e082486f6939a65c4aeaa4ea527783f?page-size=100&page-number=1`
- `findAuthorizedOperationsOrThrow()` doesn't find any operations

 However, this is wrong, because this takes the same code path as if for a `GET` of individual data. But this calls the `/history` API, and ends up with the "`PUT` and no data, so this is a 1" branch, and `authorizedOperationsForNoData()`, which returns no permissions.

So even if we wanted to honor permissions, this would be wrong.

But does this point to a bug with the `/history` API?

- [ ] check and enter `/history` API bug if needed

However, we do not want permissions to be honored, as per the decision above.

**Comment by ebruchez** [comment]
If we want the current permissions not to be honored, we need some new sort of setting to override the default behavior. This is not something we have right now. This would be the same as passing `Permissions.Undefined`, which would produce `AnyOperation`. That, or passing `Permissions.Defined()` granting `Read` to `Anyone` would work.

However, of course, we'd need to be very careful and to enable this by passing an encrypted parameter for export, which would only be passed in this case.

An external service call API would at this point not support this, and would honor permissions (but see possible `/history` but above).

**Comment by ebruchez** [comment]
In addition, the Search API also needs to support all operations so that all the data is returned.

**Comment by ebruchez** [comment]
So the idea is:

- Admin page calls export/purge API with the encrypted admin param
- export/purge API reads and then passes/forwards that parameter to the relevant persistence APIs
    - Search
    - CRUD
    - History (which currently ignores it?)

**Comment by ebruchez** [comment]
Instead of an encrypted param, now using a token with expiration.

**Comment by ebruchez** [comment]
Closing as all major items are done or new RFEs have been entered.

**Comment by ebruchez** [comment]
The idea of using a token is that we have some trusted source, in this case the Admin page, calling a service, in this case `/fr/service/persistence/export`.  We'd like that source to tell the service that it has some special permissions, in this case "I can list and read everything" (and, in the case of '/fr/service/persistence/purge', "I can list and read and delete everything"). This information has to be forwarded to the Search, CRUD, and History APIs down the line.

My first idea was to say that we just add an internal URL parameter that says "I have these rights". But I didn't want that parameter to be usable by everyone, at least not until we figure out all the security implications. It is true that, for service calls, we currently allow any authorized service caller to then impersonate any user. However, after that, permissions still kick in. A regular, non-enrypted parameter could be ok, or not. Also, services can also be called internally, for example by a form service. Someone could, by design or mistakenly, design a form that calls the CRUD API with that parameter, and gain access to other users' data, bypassing permissions.

So then I figured I'd just use an encrypted parameter, as we do for mode changes. However, in case of an error, the internal URL could be output to the logs, and someone obtaining the log could then know what encrypted parameter value to pass to gain more access, forever (or until the encryption password is changed).

So instead, I used a token like we use for the access token, which has the benefit of having an expiration time. Even if someone has access to the token through logs or otherwise, the token will expire. In this case, it expires in a matter of seconds, enough for an internal request to go through, but hopefully not enough to give time to someone to gain access if by any chance a token value was obtained. At last, this covers the case of logs being shared.

I cannot use the exact logic used by the access token, because that requires forms that enable token access. So currently, the token says "I am an admin and I can do anything with the data". We could change this to be more like the access token and list individual operations. This would allow `/export` to only request `read` and `list` permissions, for example.

One last question is the creation of the token. If someone could have access to the function that creates the token, that could be a problem. This function is internal, but we should harden it. One idea would be, for example, to whitelist pages and services where the function can be successful. Specifically:

- `/admin`
- `/fr/service/persistence/export`
- `/fr/service/persistence/purge`
